File: xaiop-sdk/nodejs/test/checkpoint.diff-isolation.test.js
import assert from "node:assert/strict";
import { describe, test } from "node:test";
import {
  DotCheckpointEngine,
  parseSync,
  SDK_VERSION,
} from "../dist/index.js";

/**
 * Diff isolation / cumulative locate — framing splits must match one-shot.
 * D1: later `>name` after `.` (synthetic object root).
 * D2: `@` into a prior-phase named array (cumulative Diff; not phase-local create {}).
 * Also: `emitDiff: false` without `onChunk` must not throw.
 */
describe("checkpoint Diff isolation (D1 + D2 + emitDiff)", () => {
  test("SDK_VERSION is 0.15.1", () => {
    assert.equal(SDK_VERSION, "0.15.1");
  });

  test("onChunk Diff mutation does not touch committedSnapshot", () => {
    const wire = ">\na:1\n.\n>\nb:2\n.\n";
    /** @type {unknown[]} */
    const chunks = [];
    const eng = new DotCheckpointEngine({
      mergeChunkWindow: false,
      onChunk: (d) => chunks.push(d),
    });
    eng.push(wire);
    eng.finish();
    assert.equal(chunks.length, 2);
    const first = /** @type {{ a: number }} */ (chunks[0]);
    first.a = 999;
    assert.deepEqual(eng.committedSnapshot, { a: 1, b: 2 });
  });

  // -------------------------------------------------------------------------
  // D1 — named enter after `.`
  // -------------------------------------------------------------------------
  describe("D1: >name after prior .", () => {
    const p1 = ">\n>meta\nname:x\n.\n";
    const p2 = ">rules-\n>\nid:R1\n<\n.\n";
    const full = p1 + p2;
    const expected = { meta: { name: "x" }, rules: [{ id: "R1" }] };

    test("parseSync baseline", () => {
      assert.deepEqual(parseSync(full), expected);
    });

    test("one push ≡ split pushes (default mergeChunkWindow)", () => {
      const one = runEngine([full], {});
      const split = runEngine([p1, p2], {});
      assert.deepEqual(one.committed, expected);
      assert.deepEqual(split.committed, expected);
      assert.deepEqual(one.committed, split.committed);
    });

    test("one push ≡ split pushes (mergeChunkWindow: false)", () => {
      const opts = { mergeChunkWindow: false };
      const one = runEngine([full], opts);
      const split = runEngine([p1, p2], opts);
      assert.deepEqual(one.committed, expected);
      assert.deepEqual(split.committed, expected);
      assert.equal(one.chunks.length, 2);
      assert.equal(split.chunks.length, 2);
      assert.deepEqual(split.chunks[0], { meta: { name: "x" } });
      assert.deepEqual(split.chunks[1], { rules: [{ id: "R1" }] });
    });

    test("char-stream framing still equals parseSync", () => {
      const eng = new DotCheckpointEngine({
        mergeChunkWindow: false,
        onChunk: () => {},
      });
      for (const ch of full) eng.push(ch);
      eng.finish();
      assert.deepEqual(eng.committedSnapshot, expected);
    });

    test("locate phase (=) still uses cumulative Diff", () => {
      const phase1 = ">\n>a\nx:1\n.\n";
      const phase2 = "=a\ny:2\n.\n";
      const wire = phase1 + phase2;
      assert.deepEqual(parseSync(wire), { a: { x: 1, y: 2 } });
      const chunks = [];
      const eng = new DotCheckpointEngine({
        mergeChunkWindow: false,
        onChunk: (d) => chunks.push(d),
      });
      eng.push(phase1);
      eng.push(phase2);
      eng.finish();
      assert.deepEqual(eng.committedSnapshot, { a: { x: 1, y: 2 } });
      assert.equal(chunks.length, 2);
      assert.deepEqual(chunks[1], { a: { x: 1, y: 2 } });
    });
  });

  // -------------------------------------------------------------------------
  // D2 — @ into prior-phase named array (researcher repro)
  // -------------------------------------------------------------------------
  describe("D2: @ into prior-phase named array", () => {
    const p0 = ">\n>orders-\n.\n";
    const p1 = "@orders\n>\na:1\n<\n.\n";
    const p2 = "@orders\n>\na:1\n<\n>\nb:2\n<\n.\n";
    const after1 = { orders: [{ a: 1 }] };
    const after2 = { orders: [{ a: 1 }, { a: 1 }, { b: 2 }] };
    const full = p0 + p1 + p2;

    test("parseSync baseline (one-shot)", () => {
      assert.deepEqual(parseSync(p0 + p1), after1);
      assert.deepEqual(parseSync(full), after2);
    });

    test("split pushes: Diff is array shape (not object); no throw", () => {
      const chunks = [];
      const eng = new DotCheckpointEngine({
        mergeChunkWindow: false,
        onChunk: (d) => chunks.push(d),
      });
      eng.push(p0);
      eng.push(p1);
      eng.push(p2);
      eng.finish();
      assert.deepEqual(eng.committedSnapshot, after2);
      assert.equal(chunks.length, 3);
      assert.deepEqual(chunks[0], { orders: [] });
      // Cumulative Diff — must see array enter, not phase-local create {}
      assert.deepEqual(chunks[1], after1);
      assert.ok(Array.isArray(chunks[1].orders));
      assert.deepEqual(chunks[2], after2);
      assert.ok(Array.isArray(chunks[2].orders));
    });

    test("one push ≡ split pushes (mergeChunkWindow on/off)", () => {
      for (const mergeChunkWindow of [true, false]) {
        const opts = { mergeChunkWindow };
        const one = runEngine([full], opts);
        const split = runEngine([p0, p1, p2], opts);
        assert.deepEqual(one.committed, after2);
        assert.deepEqual(split.committed, after2);
        assert.deepEqual(one.committed, split.committed);
      }
    });

    test("char-stream framing still equals parseSync", () => {
      const eng = new DotCheckpointEngine({
        mergeChunkWindow: false,
        onChunk: () => {},
      });
      for (const ch of full) eng.push(ch);
      eng.finish();
      assert.deepEqual(eng.committedSnapshot, after2);
    });

    test("@ create-only later phase still commits (cumulative Diff includes siblings)", () => {
      const a = ">\n>meta\nname:x\n.\n";
      const b = "@fresh\nv:1\n.\n";
      const expected = { meta: { name: "x" }, fresh: { v: 1 } };
      assert.deepEqual(parseSync(a + b), expected);
      const chunks = [];
      const eng = new DotCheckpointEngine({
        mergeChunkWindow: false,
        onChunk: (d) => chunks.push(d),
      });
      eng.push(a);
      eng.push(b);
      eng.finish();
      assert.deepEqual(eng.committedSnapshot, expected);
      assert.deepEqual(chunks[1], expected);
    });

    test("workarounds =orders / >orders- remain correct", () => {
      const base = ">\n>orders-\n.\n";
      const viaEq = base + "=orders\n>\na:1\n<\n.\n";
      const viaRe = base + ">orders-\n>\na:1\n<\n.\n";
      const expect = { orders: [{ a: 1 }] };
      assert.deepEqual(parseSync(viaEq), expect);
      assert.deepEqual(parseSync(viaRe), expect);
      for (const wire of [viaEq, viaRe]) {
        const split = wire === viaEq
          ? [base, "=orders\n>\na:1\n<\n.\n"]
          : [base, ">orders-\n>\na:1\n<\n.\n"];
        const r = runEngine(split, { mergeChunkWindow: false });
        assert.deepEqual(r.committed, expect);
        assert.ok(Array.isArray(r.chunks[1].orders));
      }
    });
  });

  // -------------------------------------------------------------------------
  // emitDiff: false without onChunk
  // -------------------------------------------------------------------------
  describe("emitDiff: false / missing onChunk", () => {
    test("emitDiff:false without onChunk does not throw", () => {
      const eng = new DotCheckpointEngine({ emitDiff: false });
      eng.push(">\na:1\n.\n");
      eng.finish();
      assert.deepEqual(eng.committedSnapshot, { a: 1 });
      assert.deepEqual(eng.snapshot, { a: 1 });
    });

    test("emitDiff:false with empty onChunk still commits", () => {
      const seen = [];
      const eng = new DotCheckpointEngine({
        emitDiff: false,
        onChunk: (d) => seen.push(d),
      });
      eng.push(">\na:1\n.\n");
      eng.finish();
      assert.deepEqual(eng.committedSnapshot, { a: 1 });
      // null Diff delivery when emitDiff off
      assert.ok(seen.length >= 1);
      assert.equal(seen[0], null);
    });

    test("missing onChunk with emitDiff:true still commits (Diff discarded)", () => {
      const eng = new DotCheckpointEngine({});
      eng.push(">\na:1\n.\n");
      eng.finish();
      assert.deepEqual(eng.committedSnapshot, { a: 1 });
    });

    test("emitDiff:false + @ array multi-phase still commits", () => {
      const eng = new DotCheckpointEngine({ emitDiff: false });
      eng.push(">\n>orders-\n.\n");
      eng.push("@orders\n>\na:1\n<\n.\n");
      eng.push("@orders\n>\nb:2\n<\n.\n");
      eng.finish();
      assert.deepEqual(eng.committedSnapshot, {
        orders: [{ a: 1 }, { b: 2 }],
      });
    });
  });
});

/**
 * @param {string[]} chunks
 * @param {object} opts
 */
function runEngine(chunks, opts) {
  /** @type {unknown[]} */
  const out = [];
  const eng = new DotCheckpointEngine({
    ...opts,
    onChunk: (d) => out.push(d),
  });
  for (const c of chunks) eng.push(c);
  eng.finish();
  return { committed: eng.committedSnapshot, chunks: out };
}

File: xaiop-sdk/java/src/test/java/io/xaiop/CheckpointDiffIsolationTest.java
package io.xaiop;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertInstanceOf;
import static org.junit.jupiter.api.Assertions.assertTrue;

import io.xaiop.stream.DotCheckpointEngine;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import org.junit.jupiter.api.Test;

/** Port of Node {@code checkpoint.diff-isolation.test.js} — D1 / D2 / emitDiff. */
class CheckpointDiffIsolationTest {

  private static final class Run {
    final Object committed;
    final List<Object> chunks;

    Run(Object committed, List<Object> chunks) {
      this.committed = committed;
      this.chunks = chunks;
    }
  }

  private static Run runEngine(List<String> chunks, boolean mergeChunkWindow) {
    List<Object> out = new ArrayList<>();
    DotCheckpointEngine eng =
        DotCheckpointEngine.Options.builder()
            .mergeChunkWindow(mergeChunkWindow)
            .onChunk(out::add)
            .build();
    for (String c : chunks) eng.push(c);
    eng.finish();
    return new Run(eng.committedSnapshot(), out);
  }

  @Test
  void onChunkDiffMutationDoesNotTouchCommitted() {
    String wire = ">\na:1\n.\n>\nb:2\n.\n";
    List<Object> chunks = new ArrayList<>();
    DotCheckpointEngine eng =
        DotCheckpointEngine.Options.builder().mergeChunkWindow(false).onChunk(chunks::add).build();
    eng.push(wire);
    eng.finish();
    assertEquals(2, chunks.size());
    @SuppressWarnings("unchecked")
    Map<String, Object> first = (Map<String, Object>) chunks.get(0);
    first.put("a", 999);
    assertEquals(Map.of("a", 1, "b", 2), eng.committedSnapshot());
  }

  @Test
  void d1NamedEnterAfterPriorDot() {
    String p1 = ">\n>meta\nname:x\n.\n";
    String p2 = ">rules-\n>\nid:R1\n<\n.\n";
    String full = p1 + p2;
    Map<String, Object> expected =
        Map.of("meta", Map.of("name", "x"), "rules", List.of(Map.of("id", "R1")));
    assertEquals(expected, Parse.parse(full));

    Run one = runEngine(List.of(full), true);
    Run split = runEngine(List.of(p1, p2), true);
    assertEquals(expected, one.committed);
    assertEquals(expected, split.committed);

    Run stepwise = runEngine(List.of(p1, p2), false);
    assertEquals(expected, stepwise.committed);
    assertEquals(2, stepwise.chunks.size());
    assertEquals(Map.of("meta", Map.of("name", "x")), stepwise.chunks.get(0));
    assertEquals(Map.of("rules", List.of(Map.of("id", "R1"))), stepwise.chunks.get(1));

    DotCheckpointEngine charStream =
        DotCheckpointEngine.Options.builder().mergeChunkWindow(false).onChunk(d -> {}).build();
    for (int i = 0; i < full.length(); i++) {
      charStream.push(String.valueOf(full.charAt(i)));
    }
    charStream.finish();
    assertEquals(expected, charStream.committedSnapshot());
  }

  @Test
  void locatePhaseUsesCumulativeDiff() {
    String phase1 = ">\n>a\nx:1\n.\n";
    String phase2 = "=a\ny:2\n.\n";
    assertEquals(Map.of("a", Map.of("x", 1, "y", 2)), Parse.parse(phase1 + phase2));
    List<Object> chunks = new ArrayList<>();
    DotCheckpointEngine eng =
        DotCheckpointEngine.Options.builder().mergeChunkWindow(false).onChunk(chunks::add).build();
    eng.push(phase1);
    eng.push(phase2);
    eng.finish();
    assertEquals(Map.of("a", Map.of("x", 1, "y", 2)), eng.committedSnapshot());
    assertEquals(2, chunks.size());
    assertEquals(Map.of("a", Map.of("x", 1, "y", 2)), chunks.get(1));
  }

  @Test
  void d2AtIntoPriorPhaseNamedArray() {
    String p0 = ">\n>orders-\n.\n";
    String p1 = "@orders\n>\na:1\n<\n.\n";
    String p2 = "@orders\n>\na:1\n<\n>\nb:2\n<\n.\n";
    Map<String, Object> after1 = Map.of("orders", List.of(Map.of("a", 1)));
    Map<String, Object> after2 =
        Map.of("orders", List.of(Map.of("a", 1), Map.of("a", 1), Map.of("b", 2)));
    String full = p0 + p1 + p2;
    assertEquals(after1, Parse.parse(p0 + p1));
    assertEquals(after2, Parse.parse(full));

    List<Object> chunks = new ArrayList<>();
    DotCheckpointEngine eng =
        DotCheckpointEngine.Options.builder().mergeChunkWindow(false).onChunk(chunks::add).build();
    eng.push(p0);
    eng.push(p1);
    eng.push(p2);
    eng.finish();
    assertEquals(after2, eng.committedSnapshot());
    assertEquals(3, chunks.size());
    assertEquals(Map.of("orders", List.of()), chunks.get(0));
    assertEquals(after1, chunks.get(1));
    assertInstanceOf(List.class, ((Map<?, ?>) chunks.get(1)).get("orders"));
    assertEquals(after2, chunks.get(2));

    for (boolean merge : new boolean[] {true, false}) {
      Run one = runEngine(List.of(full), merge);
      Run split = runEngine(List.of(p0, p1, p2), merge);
      assertEquals(after2, one.committed);
      assertEquals(after2, split.committed);
    }
  }

  @Test
  void atCreateOnlyLaterPhaseStillCommits() {
    String a = ">\n>meta\nname:x\n.\n";
    String b = "@fresh\nv:1\n.\n";
    Map<String, Object> expected = Map.of("meta", Map.of("name", "x"), "fresh", Map.of("v", 1));
    assertEquals(expected, Parse.parse(a + b));
    List<Object> chunks = new ArrayList<>();
    DotCheckpointEngine eng =
        DotCheckpointEngine.Options.builder().mergeChunkWindow(false).onChunk(chunks::add).build();
    eng.push(a);
    eng.push(b);
    eng.finish();
    assertEquals(expected, eng.committedSnapshot());
    assertEquals(expected, chunks.get(1));
  }

  @Test
  void emitDiffFalseWithoutOnChunkDoesNotThrow() {
    DotCheckpointEngine eng = DotCheckpointEngine.Options.builder().emitDiff(false).build();
    eng.push(">\na:1\n.\n");
    eng.finish();
    assertEquals(Map.of("a", 1), eng.committedSnapshot());
    assertEquals(Map.of("a", 1), eng.snapshot());
  }

  @Test
  void missingOnChunkStillCommits() {
    DotCheckpointEngine eng = DotCheckpointEngine.Options.builder().build();
    eng.push(">\na:1\n.\n");
    eng.finish();
    assertEquals(Map.of("a", 1), eng.committedSnapshot());
  }

  @Test
  void emitDiffFalseAtArrayMultiPhase() {
    DotCheckpointEngine eng = DotCheckpointEngine.Options.builder().emitDiff(false).build();
    eng.push(">\n>orders-\n.\n");
    eng.push("@orders\n>\na:1\n<\n.\n");
    eng.push("@orders\n>\nb:2\n<\n.\n");
    eng.finish();
    assertEquals(
        Map.of("orders", List.of(Map.of("a", 1), Map.of("b", 2))), eng.committedSnapshot());
  }

  @Test
  void workaroundsRemainCorrect() {
    String base = ">\n>orders-\n.\n";
    Map<String, Object> expect = Map.of("orders", List.of(Map.of("a", 1)));
    Run viaEq = runEngine(List.of(base, "=orders\n>\na:1\n<\n.\n"), false);
    Run viaRe = runEngine(List.of(base, ">orders-\n>\na:1\n<\n.\n"), false);
    assertEquals(expect, viaEq.committed);
    assertEquals(expect, viaRe.committed);
    assertTrue(((Map<?, ?>) viaEq.chunks.get(1)).get("orders") instanceof List);
  }
}
